### PR TITLE
Remove the coordinator

### LINKED ISSUE #1744: Refactor and rethink the coordinator/monitor implementation
When an elastic policy is dispatched to fleet server, one of the fleet server will be elected and take take the control of some of the management task, like when to unenroll agent after a last seen timeout. I've fixed a race condition issue discovered in #1738. When I've jumped the code I found it a bit harder to ready to really understand the flow of events and when goroutine were created and removed. 

We need to evaluate if we need to add logic to this area of the code, if this is the case we should really invest some time in refactoring the logic, here a few things to consider changing:

- We create a goroutine per agent policy, as the number of policy is low this is perfectly fine, but I think that logic could be handled by a single cleanup event loop in a single goroutine.
- We are exposing internal fields from monitorT object in the test suite, we should hide all access to the internal field using accessor even if this is only for testing. This allow a single locking logic.
- Looking at the code, It look like possible the usage of multiple internal fields into a watcher struct that would encapsulate more logic.
- Internal state of the monitor bleeds into the goroutine execution, this make it harder to lock or prevent concurrent access to the resource. Encapsulating that logic into his own object would make it simple to test and verifies.

### PR DESCRIPTION
**Pull request by michel-laterman**
## What is the problem this PR solves?

fleet-server coordinator and leader election mechanisms are nops that add unneeded complexity to understanding the codebase.

## How does this PR solve the problem?

Remove the policy coordinator and policy leader election mechanisms from fleet-server. Deprecate the coordinator_idx value in fleet-server's json schema and remove coordinator_idx references when processing policies.

## Design Checklist

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- [ ] I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.
- ~~I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.~~

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

## Related Issues

- Closes #1744

### CODE REVIEW

File: internal/pkg/dl/migration.go [hunk 1]
```diff
@@ -274,16 +274,17 @@ ctx._source.policy_output_permissions_hash=null;
...
 
-// migratePolicyCoordinatorIdx increases the policy's CoordinatorIdx to force
+// migratePolicyRevisionIdx increases the policy's RevisionIdx to force
 // a policy update ensuring the output data will be migrated to the new
 // Agent.Outputs field. See migrateAgentOutputs and https://github.com/elastic/fleet-server/issues/1672
 // for details.
-func migratePolicyCoordinatorIdx() (string, string, []byte, error) {
-	const migrationName = "PolicyCoordinatorIdx"
+// NOTE(michel-laterman): Updated to increment the RevisionIdx instead of the CoordinatorrIdx as the coordinator is being removed.
+func migratePolicyRevisionIdx() (string, string, []byte, error) {
```
[the remark this answers is not in the document]
Comment (kpollich) [reply internal/pkg/dl/migration.go:282]: Bumping the revision is the same mechanism Fleet uses when outputs or Fleet Server hosts change in order to force a policy redeploy, so this sounds reasonable to me.

File: internal/pkg/dl/constants.go [hunk 2]
```diff
@@ -14,8 +14,6 @@ const (
 	FleetArtifacts         = ".fleet-artifacts"
 	FleetEnrollmentAPIKeys = ".fleet-enrollment-api-keys"
 	FleetPolicies          = ".fleet-policies"
-	FleetPoliciesLeader    = ".fleet-policies-leader"
-	FleetServers           = ".fleet-servers"
```
[the remark this answers is not in the document]
Comment (juliaElastic) [reply internal/pkg/dl/constants.go:18]: we should create a follow up issue to check and clean up this index in the next major if not used

**Review by juliaElastic — APPROVED** [review]
Changes LGTM.
Is there anything that can go wrong if there are multiple fleet servers, some of them on older version with a coordinator?

**Comment by michel-laterman** [comment]
If we have the scenario where an older version is running at the same time as a newer version, it will only provide policies where the `coordinator_idx` has been updated to >0, however a newer version may serve (the same policy) with a 0 value.
If an agent gets it's policy from a new fleet-server, then checks in with an older instance it may immediately get a policy change action with the `coordinator_idx` value changed.
The rest of the policy data will not change.

@cmacknz, would this sequence present an issue to the agent?

**Comment by cmacknz** [comment]
Just to confirm this only duplicates the policy change action and not other action types?

If the policy is exactly the same I *think* this is fine, the agent should realize there are no changes to the running set of components and make not changes. I would recommend testing this quickly to confirm though.

There is no actual requirement for independent Fleet Servers to run the same version is there? In practice this is often the case but there is nothing enforcing.

I am wondering if we could get into an edge case where the agent is continuously re-processing a policy change going between two Fleet servers with different versions. Is this possible or something that we need to guard against? Does Fleet server itself tolerate the `coordinator_idx` constantly toggling?

**Comment by michel-laterman** [comment]
OK, I did some testing and removing the coordinator_idx value should not effect the agent.

I tested the following sequence
1. agent enrolls in fleet-server with coordinator
2. fleet-server coordinator is removed
3. policy is updated
4. fleet-server with coordinator replaces fleet-server instance

The agent behaved normally throughout.
I also tried the sequence with a 2nd fleet-server instance (with a coordinator) that did not serve the agent (but updated the polices that the agent uses), it did not effect the other fleet-server instance.

File: internal/pkg/dl/constants.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nchaulet) [reply internal/pkg/dl/constants.go:18]: @michel-laterman the Fleet status API rely on the `.fleet-servers` entries https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/routes/setup/handlers.ts#L22 so we should change that in Kibana before going forward with that PR

[the remark this answers is not in the document]
Comment (michel-laterman) [reply internal/pkg/dl/constants.go:18]: Do we want to change Kibana's behaviour as well, or change fleet-server to register itself somewhere else?

File: internal/pkg/dl/migration.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nchaulet) [reply internal/pkg/dl/migration.go:282]: Currently the source of truth for that revision_idx in kibana is in the agent policy saved object, so kibana could try to push policies with policy_revision_idx already used

File: internal/pkg/dl/constants.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nchaulet) [reply internal/pkg/dl/constants.go:18]: I think for this one it's probably okay to change the Kibana behavior, we can rely on retrieving agents with fleet-server installed instead.

**Comment by michel-laterman** [comment]
We need to hold off on merging this until the following Kibana issues are resolved:
* https://github.com/elastic/kibana/issues/173537
* https://github.com/elastic/kibana/issues/173538

**Comment by jlind23** [comment]
@michel-laterman does this one also closes - https://github.com/elastic/kibana/issues/173538

File: internal/pkg/server/fleet_integration_test.go [hunk 3]
```diff
@@ -1370,3 +1371,148 @@ func Test_SmokeTest_CheckinPollShutdown(t *testing.T) {
...
+	t.Logf("Attempting to remove api_key attribute from: %s, body: %s", id, string(p))
+	err = srv.bulker.Update(
+		ctx,
+		dl.FleetAgents,
+		id,
+		p,
+		bulk.WithRefresh(),
+		bulk.WithRetryOnConflict(3),
+	)
+	require.NoError(t, err)
```
Comment (michel-laterman) [inline internal/pkg/server/fleet_integration_test.go:1487]: This was failing when I ran the integration tests locally with
```
    fleet_integration_test.go:1478: Attempting to remove api_key attribute from: 8abbbaf3-cfec-41ae-a086-540cc8604885, body: {"script":{"lang": "painless", "source": "ctx._source['outputs'][params.output].api_key == ''; ctx._source['outputs'][params.output].api_key_id == '';", "params": {"output": "default"}}}
    fleet_integration_test.go:1487:
        	Error Trace:	/Users/laterman/git/fleet-server/internal/pkg/server/fleet_integration_test.go:1487
        	Error:      	Received unexpected error:
        	            	elastic fail 400: illegal_argument_exception: failed to execute script: script_exception: compile error
        	Test:       	Test_SmokeTest_Verify_v85Migrate
--- FAIL: Test_SmokeTest_Verify_v85Migrate (3.98s)
```
I may need some help figuring out what the actual error is

**Comment by jen-huang** [comment]
> We need to hold off on merging this until the following Kibana issues are resolved:
> 
> * [[Fleet] Remove dependency on `.fleet-servers` index kibana#173537](https://github.com/elastic/kibana/issues/173537)
> * [[Fleet] policy revision_idx source of truth kibana#173538](https://github.com/elastic/kibana/issues/173538)

FYI https://github.com/elastic/kibana/issues/173537 was recently completed and merged, so this should be unblocked now. Are there any BWC scenarios we need to consider?

**Comment by michel-laterman** [comment]
> FYI [elastic/kibana#173537](https://github.com/elastic/kibana/issues/173537) was recently completed and merged, so this should be unblocked now. Are there any BWC scenarios we need to consider?

I can't think of any.
We can restest the sequence from https://github.com/elastic/fleet-server/pull/3131#issuecomment-1854866235 with the fleet-ui changesto make sure everything works as expected

**Comment by AndersonQ** [comment]
@pierrehilbert I believe the control-plane and data-plane team aren't up-to-date yet. Perhaps you'd like to change who is supposed to review this PR on behalf of the control-plane.

**Comment by michel-laterman** [comment]
buildkite run perf-tests

**Comment by michel-laterman** [comment]
I've re-validated this change;
I ran ES + Kibana locally then:

1) started a stand-alone fleet-server (with coordinator)
2) enrolled an agent
3) halted fleet-server and waited until agent reported offile
4) started stand-alone fleet-server (no coordinator)
5) issued a policy change
6) observed agent received update and was healthy

At this point I'm certain that the change will function.

However I continued the experiment and once more halted the fleet-server and replaced it with one with the coordinator and saw the agent was still healthy.

I'm confident with the change and testing done @nchaulet are there any other changes we'll need from fleet in Kibana, seeing how we've closed the issues in https://github.com/elastic/fleet-server/pull/3131#issuecomment-1860796331; opr are we OK to merge?

**Comment by jen-huang** [comment]
@michel-laterman It would be good to test an existing fleet server (with coordinator) + agents, and upgrading fleet server to this version without coordinator.

**Comment by michel-laterman** [comment]
@jen-huang, that's what I tested

**Comment by jen-huang** [comment]
🤦 I misread some of the lines. Ty :)

**Review by nchaulet — APPROVED** [review]
LGTM 🚀  
note this will be deployed to serverless as soon the different quality gates does the migration from that version back to using the coordinator will work in case we need to revert? (I think so but just checking)

**Comment by michel-laterman** [comment]
I did a brief test where I stopped a fleet-server (with no coordinator) and replaced it with one with a coordinator and the agent was able to become healthy in the UI

**Pull request merged**